server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileProjectsAction.java

```diff
@@ -0,0 +1,164 @@
+      .setRequired(true)
+      .setExampleValue("sonar-way-java-12345");
+    projects.createParam(PARAM_SELECTED)
+      .setDescription("If specified, return only selected or deselected projects.")
+      .setPossibleValues(SELECTION_SELECTED, SELECTION_DESELECTED, SELECTION_ALL)
+      .setDefaultValue(SELECTION_ALL);
+    projects.createParam(PARAM_QUERY)
+      .setDescription("If specified, return only projects whose name match the query.");
+    projects.createParam(PARAM_PAGE_SIZE)
+      .setDescription("Size for the paging to apply").setDefaultValue(100);
+    projects.createParam(PARAM_PAGE)
+      .setDescription("Index of the page to display").setDefaultValue(1);
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    String profileKey = request.mandatoryParam(PARAM_KEY);
+
+    DbSession session = dbClient.openSession(false);
+
+    try {
+      checkProfileExists(profileKey, session);
+      String selected = request.param(PARAM_SELECTED);
+      String query = request.param(PARAM_QUERY);
+      int pageSize = request.mandatoryParamAsInt(PARAM_PAGE_SIZE);
+      int page = request.mandatoryParamAsInt(PARAM_PAGE);
+
+      List<ProjectQprofileAssociationDto> projects = loadProjects(profileKey, session, selected, query);
+
+      Collection<Long> projectIds = Collections2.transform(projects, new NonNullInputFunction<ProjectQprofileAssociationDto, Long>() {
+        @Override
+        protected Long doApply(ProjectQprofileAssociationDto input) {
+          return input.getProjectId();
+        }
+      });
+
+      final Collection<Long> authorizedProjectIds = dbClient.authorizationDao().keepAuthorizedProjectIds(session, projectIds, UserSession.get().userId(), UserRole.USER);
+      Iterable<ProjectQprofileAssociationDto> authorizedProjects = Iterables.filter(projects, new Predicate<ProjectQprofileAssociationDto>() {
+        @Override
+        public boolean apply(ProjectQprofileAssociationDto input) {
+          return authorizedProjectIds.contains(input.getProjectId());
+        }
+      });
+
+      Paging paging = Paging.create(pageSize, page, authorizedProjectIds.size());
+
+      List<ProjectQprofileAssociationDto> pagedAuthorizedProjects = Lists.newArrayList(authorizedProjects);
+      if (pagedAuthorizedProjects.size() <= paging.offset()) {
+        pagedAuthorizedProjects = Lists.newArrayList();
+      } else if (pagedAuthorizedProjects.size() > paging.pageSize()) {
+        pagedAuthorizedProjects = pagedAuthorizedProjects.subList(paging.offset(), paging.offset() + pageSize);
+      }
+
+      writeProjects(response.newJsonWriter(), pagedAuthorizedProjects, paging);
+    } finally {
+      session.close();
+    }
+  }
+
+  private void checkProfileExists(String profileKey, DbSession session) {
+    try {
+      dbClient.qualityProfileDao().getNonNullByKey(session, profileKey);
+    } catch (IllegalArgumentException doesNotExist) {
```

Comment:
I do not like catching an exception is this case. I think it would be better (and more efficient) to use dbClient.qualityProfileDao().getByKey(), and return NotFoundException(String.format("Could not find a quality profile with key '%s'", profileKey)) if the result is null.